odfuzz/fuzzer.py

```diff
@@ -1221,13 +1221,23 @@ def build_filter_string(filter_data):
 
 def replace_forbidden_characters(parts):
     for data in parts:
-        if data['operand'].startswith('\''):
-            data['operand'] = '\'' + replace_special_characters(data['operand'][1:-1]) + '\''
         if 'params' in data:
+            # replace special characters within function's parameters
             if data['params']:
                 for index, param in enumerate(data['params']):
-                    if param.startswith('\''):
-                        data['params'][index] = '\'' + replace_special_characters(param[1:-1]) + '\''
+                    replace_part_characters(data['params'], index)
+        else:
+            replace_part_characters(data, 'operand')
+
+
+def replace_part_characters(data, key):
```

Comment:
One question:

Why is replace_forbidden_characters() needed?  If edm_double() generator produces something, which is not usable for URL part, shouldn't solving its fault be under its boundary (method, or at least generators module)?

This question is related to testability - no generator method has coverage, but if I would add test to cover edm_double(), it still be kind of useless test, am I correct? not covering everything what is expecting to be "URL part string of double number).

I understand that this is outside of the scope of this pull request and that as it is now, it will fix the issue #16 - so I am inclined to approve this. This comment is for discussion purpose and I guess new issue will be created if agreed on.

Reply:
Great questions. For testing purposes, we would probably be able to do something similar to this:
https://github.com/SAP/odfuzz/blob/e2db91f8c4cf9e42066079b2e04ec5f80d3edfd7/odfuzz/mutators.py#L202-L203
https://github.com/SAP/odfuzz/blob/e2db91f8c4cf9e42066079b2e04ec5f80d3edfd7/tests/test_mutator.py#L31-L33

On the other hand, you have correctly mentioned that it could be useless to implement regular unit tests for this scenario. I will take a closer look at Hypothesis testing library later this day.

> Why is replace_forbidden_characters() needed?

This might be one of the design flaws of ODfuzz. The generator generates raw strings which are not URL encoded. An output of the generator is correct because it yields data in a valid format (see the section "Primitive Data Types" - https://www.odata.org/documentation/odata-version-2-0/overview/). In order to send an HTTP request, it is necessary to encode that data. The generator for sure does not know what is going to happen with the generated data. Due to that, the data are not encoded in the generator itself.

In our case, we utilize the generator only to generate data which are going to be dispatched via HTTP as soon as possible. So, the generator could sanitize the output in every iteration.